x-pack/plugins/fleet/server/services/epm/packages/cleanup.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { SavedObjectsClientContract } from 'src/core/server';
+
+import { removeArchiveEntries } from '../archive/storage';
+
+import { ASSETS_SAVED_OBJECT_TYPE, PACKAGE_POLICY_SAVED_OBJECT_TYPE } from '../../../../common';
+import type { PackageAssetReference } from '../../../../common';
+import { packagePolicyService } from '../../package_policy';
+import { appContextService } from '../..';
+
+export async function removeOldAssets(options: {
+  savedObjectsClient: SavedObjectsClientContract;
+  pkgName: string;
+  oldVersion: string;
+}) {
+  const { savedObjectsClient, pkgName, oldVersion } = options;
+
+  // check if old version has assets
+  const assets = await savedObjectsClient.find<PackageAssetReference>({
+    type: ASSETS_SAVED_OBJECT_TYPE,
+    filter: `${ASSETS_SAVED_OBJECT_TYPE}.attributes.package_name:${pkgName} AND ${ASSETS_SAVED_OBJECT_TYPE}.attributes.package_version:${oldVersion}`,
+    // how to query if more than one page?
+    perPage: 100,
+    fields: ['id'],
+  });
+  const refs = assets.saved_objects.map(
+    (obj) => ({ id: obj.id, type: ASSETS_SAVED_OBJECT_TYPE } as PackageAssetReference)
+  );
+
+  // check if any policies are using this package version
+  const { total } = await packagePolicyService.list(savedObjectsClient, {
+    kuery: `${PACKAGE_POLICY_SAVED_OBJECT_TYPE}.package.name:${pkgName} AND ${PACKAGE_POLICY_SAVED_OBJECT_TYPE}.package.version:${oldVersion}`,
+    page: 0,
+    perPage: 0,
```

Comment:
Will `perPage: 0` ever return a total more than 0?

Also maybe we should do this check first before paging through all the refs and deleting them. That way we can delete the refs in batches without requiring that we page through all and load them into memory before deleting.

Reply:
yes, I took this code from remove.ts where the same check is used, also tested and it returned > 0 if there were policies
good point about the order, will do the check first